File: src/pkjs/solar.js
// Solar event calculations (NOAA algorithm, center-of-disk geometric elevation).
//
// This runs on the PHONE, not the watch: the watch has a hard XS memory
// ceiling and this is the largest block of arithmetic in the project. The
// phone computes a rolling window of upcoming events and sends the
// timestamps, so the watch stays correct offline for as long as the window
// lasts (see src/pkjs/index.js).
//
// CommonJS so both pkjs and the node test suite can load it.

const DEG = Math.PI / 180;
const MS_PER_DAY = 86400000;

// Standard elevation angles (degrees relative to geometric horizon, center of disk).
const SUNRISE_SET_ANGLE = -0.833; // refraction 34' + semidiameter 16'
const TZEIT_ANGLE = -8.5;

function julianCentury(ms) {
	return (ms / MS_PER_DAY + 2440587.5 - 2451545) / 36525;
}

// Solar declination (deg) and equation of time (minutes) at Julian century T.
function solarParams(T) {
	const L0 = (280.46646 + T * (36000.76983 + 0.0003032 * T)) % 360;
	const M = 357.52911 + T * (35999.05029 - 0.0001537 * T);
	const e = 0.016708634 - T * (0.000042037 + 0.0000001267 * T);
	const C =
		Math.sin(M * DEG) * (1.914602 - T * (0.004817 + 0.000014 * T)) +
		Math.sin(2 * M * DEG) * (0.019993 - 0.000101 * T) +
		Math.sin(3 * M * DEG) * 0.000289;
	const omega = 125.04 - 1934.136 * T;
	const appLong = L0 + C - 0.00569 - 0.00478 * Math.sin(omega * DEG);
	const meanObliq =
		23 + (26 + (21.448 - T * (46.815 + T * (0.00059 - T * 0.001813))) / 60) / 60;
	const obliq = meanObliq + 0.00256 * Math.cos(omega * DEG);
	const decl = Math.asin(Math.sin(obliq * DEG) * Math.sin(appLong * DEG)) / DEG;
	const halfTan = Math.tan((obliq / 2) * DEG);
	const y = halfTan * halfTan;   // ** is too new for the pkjs bundler's parser
	const eqTime =
		(4 / DEG) *
		(y * Math.sin(2 * L0 * DEG) -
			2 * e * Math.sin(M * DEG) +
			4 * e * y * Math.sin(M * DEG) * Math.cos(2 * L0 * DEG) -
			0.5 * y * y * Math.sin(4 * L0 * DEG) -
			1.25 * e * e * Math.sin(2 * M * DEG));
	return { decl, eqTime };
}

// Hour angle (deg) at which the sun's center reaches elevation angleDeg; null if never.
function hourAngle(latDeg, declDeg, angleDeg) {
	const cosHA =
		(Math.cos((90 - angleDeg) * DEG) -
			Math.sin(latDeg * DEG) * Math.sin(declDeg * DEG)) /
		(Math.cos(latDeg * DEG) * Math.cos(declDeg * DEG));
	if (cosHA < -1 || cosHA > 1) return null;
	return Math.acos(cosHA) / DEG;
}

// Rising (rising=true) or setting event on the UTC day containing dayMs.
// Refined by re-evaluating declination/EoT at the event estimate.
function eventForUTCDay(dayMs, lat, lon, angleDeg, rising) {
	const dayStart = Math.floor(dayMs / MS_PER_DAY) * MS_PER_DAY;
	let guess = dayStart + MS_PER_DAY / 2;
	for (let i = 0; i < 3; i++) {
		const { decl, eqTime } = solarParams(julianCentury(guess));
		const ha = hourAngle(lat, decl, angleDeg);
		if (ha === null) return null;
		const noonMin = 720 - 4 * lon - eqTime;
		const evMin = rising ? noonMin - 4 * ha : noonMin + 4 * ha;
		guess = dayStart + evMin * 60000;
	}
	return Math.round(guess);
}

// All rise/set events at angleDeg over the three UTC days around nowMs, sorted.
function sunEvents(nowMs, lat, lon, angleDeg = SUNRISE_SET_ANGLE) {
	const evs = [];
	for (let d = -1; d <= 1; d++) {
		for (const rising of [true, false]) {
			const t = eventForUTCDay(nowMs + d * MS_PER_DAY, lat, lon, angleDeg, rising);
			if (t !== null) evs.push({ t, rising });
		}
	}
	evs.sort((a, b) => a.t - b.t);
	return evs;
}

// The current half-day: previous sunrise/sunset to next. isDay = sun currently up.
// Returns null at polar latitudes when the sun doesn't cross the horizon.
function bracket(nowMs, lat, lon) {
	const evs = sunEvents(nowMs, lat, lon);
	let prev = null;
	let next = null;
	for (const ev of evs) {
		if (ev.t <= nowMs) prev = ev;
		else {
			next = ev;
			break;
		}
	}
	if (!prev || !next) return null;
	return { start: prev.t, end: next.t, isDay: prev.rising };
}

// Next rise/set event at angleDeg strictly after afterMs; null if none within a day.
function nextEvent(afterMs, lat, lon, angleDeg, rising) {
	const evs = sunEvents(afterMs, lat, lon, angleDeg);
	for (const ev of evs) {
		if (ev.rising === rising && ev.t > afterMs) return ev.t;
	}
	return null;
}

module.exports = { sunEvents, bracket, nextEvent, SUNRISE_SET_ANGLE, TZEIT_ANGLE, MS_PER_DAY };

File: src/c/solar.c
#include "solar.h"

#include <math.h>
#include <stddef.h>

// The Pebble toolchain compiles without the GNU extensions that define M_PI,
// so spell it out rather than relying on math.h.
#define PI 3.14159265358979323846
#define DEG (PI / 180.0)

static double julian_century(double ms) {
  return (ms / MS_PER_DAY + 2440587.5 - 2451545.0) / 36525.0;
}

// Solar declination (deg) and equation of time (minutes) at Julian century T.
static void solar_params(double T, double *decl, double *eq_time) {
  double L0 = fmod(280.46646 + T * (36000.76983 + 0.0003032 * T), 360.0);
  double M = 357.52911 + T * (35999.05029 - 0.0001537 * T);
  double e = 0.016708634 - T * (0.000042037 + 0.0000001267 * T);
  double C = sin(M * DEG) * (1.914602 - T * (0.004817 + 0.000014 * T)) +
             sin(2 * M * DEG) * (0.019993 - 0.000101 * T) +
             sin(3 * M * DEG) * 0.000289;
  double omega = 125.04 - 1934.136 * T;
  double app_long = L0 + C - 0.00569 - 0.00478 * sin(omega * DEG);
  double mean_obliq =
      23.0 + (26.0 + (21.448 - T * (46.815 + T * (0.00059 - T * 0.001813))) / 60.0) / 60.0;
  double obliq = mean_obliq + 0.00256 * cos(omega * DEG);
  double half_tan = tan((obliq / 2.0) * DEG);
  double y = half_tan * half_tan;

  *decl = asin(sin(obliq * DEG) * sin(app_long * DEG)) / DEG;
  *eq_time = (4.0 / DEG) *
             (y * sin(2 * L0 * DEG) - 2 * e * sin(M * DEG) +
              4 * e * y * sin(M * DEG) * cos(2 * L0 * DEG) -
              0.5 * y * y * sin(4 * L0 * DEG) - 1.25 * e * e * sin(2 * M * DEG));
}

// Hour angle (deg) at which the sun's centre reaches angle_deg.
// False when it never does.
static bool hour_angle(double lat_deg, double decl_deg, double angle_deg, double *out) {
  double cos_ha = (cos((90.0 - angle_deg) * DEG) -
                   sin(lat_deg * DEG) * sin(decl_deg * DEG)) /
                  (cos(lat_deg * DEG) * cos(decl_deg * DEG));
  if (cos_ha < -1.0 || cos_ha > 1.0) return false;
  *out = acos(cos_ha) / DEG;
  return true;
}

// Rising or setting event on the UTC day containing day_ms, refined by
// re-evaluating declination and equation of time at the event estimate.
static bool event_for_utc_day(double day_ms, double lat, double lon,
                              double angle_deg, bool rising, double *out) {
  double day_start = floor(day_ms / MS_PER_DAY) * MS_PER_DAY;
  double guess = day_start + MS_PER_DAY / 2.0;
  for (int i = 0; i < 3; i++) {
    double decl, eq_time, ha;
    solar_params(julian_century(guess), &decl, &eq_time);
    if (!hour_angle(lat, decl, angle_deg, &ha)) return false;
    double noon_min = 720.0 - 4.0 * lon - eq_time;
    double ev_min = rising ? noon_min - 4.0 * ha : noon_min + 4.0 * ha;
    guess = day_start + ev_min * 60000.0;
  }
  *out = round(guess);
  return true;
}

int solar_events(double now_ms, double lat, double lon, double angle_deg,
                 SolarEvent *out) {
  int n = 0;
  for (int d = -1; d <= 1; d++) {
    for (int r = 1; r >= 0; r--) {  // rising first, matching the JS source order
      double t;
      if (event_for_utc_day(now_ms + d * MS_PER_DAY, lat, lon, angle_deg,
                            r == 1, &t)) {
        out[n].t_ms = t;
        out[n].rising = (r == 1);
        n++;
      }
    }
  }
  // Insertion sort by time; n is at most SOLAR_MAX_EVENTS.
  for (int i = 1; i < n; i++) {
    SolarEvent key = out[i];
    int j = i - 1;
    while (j >= 0 && out[j].t_ms > key.t_ms) {
      out[j + 1] = out[j];
      j--;
    }
    out[j + 1] = key;
  }
  return n;
}

SolarBracket solar_bracket(double now_ms, double lat, double lon) {
  SolarBracket b = {false, 0, 0, false};
  SolarEvent evs[SOLAR_MAX_EVENTS];
  int n = solar_events(now_ms, lat, lon, SUNRISE_SET_ANGLE, evs);

  const SolarEvent *prev = NULL;
  const SolarEvent *next = NULL;
  for (int i = 0; i < n; i++) {
    if (evs[i].t_ms <= now_ms) {
      prev = &evs[i];
    } else {
      next = &evs[i];
      break;
    }
  }
  if (!prev || !next) return b;

  b.valid = true;
  b.start_ms = prev->t_ms;
  b.end_ms = next->t_ms;
  b.is_day = prev->rising;
  return b;
}

bool solar_next_event(double after_ms, double lat, double lon, double angle_deg,
                      bool rising, double *out_ms) {
  SolarEvent evs[SOLAR_MAX_EVENTS];
  int n = solar_events(after_ms, lat, lon, angle_deg, evs);
  for (int i = 0; i < n; i++) {
    if (evs[i].rising == rising && evs[i].t_ms > after_ms) {
      *out_ms = evs[i].t_ms;
      return true;
    }
  }
  return false;
}
